### ISSUE TITLE
Tracking Issue for `ascii::Char` (ACP 179)

### ISSUE THREAD

**Issue by scottmcm** [rust-lang/rust#110998]
Feature gate: `#![feature(ascii_char)]` `#![feature(ascii_char_variants)]`

This is a tracking issue for the `ascii::Char` type from https://github.com/rust-lang/libs-team/issues/179

### Public API

```rust
// core::ascii

struct Char(u8 is 0..=127);

impl Debug for Char { … }
impl Display for Char { … }

impl Char {
    const fn new(c: char) -> Option<Self> { … }
    const fn from_u8(x: u8) -> Option<Self> { … }
    const unsafe fn from_u8_unchecked(x: u8) -> Self { … }
}

impl From<Char> for u8 { … }
impl From<Char> for char { … }
impl From<&[Char]> for &str { … }
```

### Steps / History

- [ ] Implementation: #...
- [ ] Final comment period (FCP)[^1]
- [ ] Stabilization PR

### Unresolved Questions

- What should it be named?
- How should its `Debug` impl work?

[^1]: https://std-dev-guide.rust-lang.org/feature-lifecycle/stabilization.html

**Comment by ChayimFriedman2** [comment]
Do we need both `u8::as_ascii()` and `ascii::Char::from_u8()`?

**Comment by scottmcm** [comment]
Regarding the enum, @kupiakos said the following in 

> The names or properties of ASCII characters will never change and there aren't many of them, and so we might as well expose them as variants that can participate in pattern matching.

And mentioned icu4x's enum,

https://github.com/unicode-org/icu4x/blob/b6c4018a736e79790898c5b91ff3ab25d33192c2/utils/tinystr/src/asciibyte.rs#L8-L11

Though given those names it's possible it's just doing that for the niche and not for direct use.

(Man I wish we had private variants as a possibility.)

To your questions, @BurntSushi ,

1. Yeah, there's at least three other possibilites here: a) put the enum in a newtype (like [`ptr::Alignment`](https://doc.rust-lang.org/nightly/std/ptr/struct.Alignment.html)) to still get the niches without exposing the variants b) use the rustc-internal magic c) just don't get a niche for now since that's not needed for its core scenario of convertibility to UTF-8 (though it's certainly nice to have)

2. Given that we're overall not fond of `as` casts, and it also allows weirder things like `as i64`, I actually wish the cast could be disabled, and have people use `as_u8` as you mention.  But that might also just happen by deprecating or warning on these casts, and eventually having `ascii::Char: AsRepr<Repr = u8>` does seem reasonable.

3. Now that #105076 has landed, I've been thinking of redoing the internals of those using `ascii::Char` to try out how it feels using the names.  That would mean code like

```rust
            '\0' => EscapeDebug::backslash(Null),
            '\r' => EscapeDebug::backslash(CarriageReturn),
            '\n' => EscapeDebug::backslash(LineFeed ),
```

because without the variants it ends up being more like

```rust
            '\0' => EscapeDebug::backslash(b'\0'.as_ascii().unwrap()),
            '\r' => EscapeDebug::backslash(b'\r'.as_ascii().unwrap()),
            '\n' => EscapeDebug::backslash(b'\n'.as_ascii().unwrap()),
```

which is nice in that it keeps the escape sequences lining up, but having to copy-paste `.as_ascii().unwrap()` all over like that doesn't fill me with joy.

I guess this is yet another situation in which I'd like custom literals.  Or something like `a'\n'`, I guess, but that's *way* too far a jump to propose right now.

**Comment by clarfonthey** [comment]
I wasn't in the discussion when this type was created (which I fully support), but is there a reason to explicitly use an enum instead of a newtype + internal niche attribute like `NonZeroU8`?

If we decide to go the pattern types route (#107606) then `NonZeroU8` could become `u8 is 1..` and this could become `u8 as ..128`. Alternatively, if we go the generic integers route (#2581) then this could become `uint<7>` or just `u7`.

IMHO, explicitly going with an enum kind of prevents these kinds of optimisations later down the road, since it forces an enum as the representation.

**Comment by BurntSushi** [comment]
@clarfonthey I think the discussion above answers your question about why we went with the enum for now. In summary, I think its principle upside is that it gives nice names to each character. Not particularly compelling, but nice.

My general feeling is that nobody is stuck on using an enum here. My best guess is that unless something compelling comes up, I wouldn't be surprised if we changed to an opaque representation such that we could change to an enum later in a compatible manner. (I think that's possible?) Purely because it's the conservative route.

I don't grok the second two paragraphs of your comment. Probably because I don't know what "pattern types" are. I also don't know what `uint<7>` means. (Don't have time to follow your links at the moment.)

**Comment by clarfonthey** [comment]
I mean, having nice names is still achievable with constants, and this still would make match expressions work as you'd expect. Personally, the benefits I'm seeing here are the ability to have more-performant code without the need for extra unsafe code, since any slice of ASCII characters is automatically valid UTF-8. Otherwise, a type invariant would mostly be unnecessary, and you could just have constants for the names.

Explaining the two proposals:

* Pattern types just allow you to write `T is pat` which is a new type that means `T, but only values which match pat`. So, `Option<T> is Some(_)` means "options which are some" and `u8 is 1..` is effectively `NonZeroU8`.
* Generic integers is a separate proposal which "converts" all the existing `uN` and `iN` types into aliases for a more generic `uint<N>` and `int<N>` type, which has the benefit of allowing non-power-of-two sizes and sizes larger than 128. The idea behind `uint<7>` is that it would be equivalent to `u8` internally with the first bit always zero, although still allow all the normal operations on integers like pattern-matching and be castable to `u8`.

The pattern type version is very enticing because it very naturally allows exhaustive matching, while still just being a compile-time constraint. And like I said, we could still have constants for the different names for characters without really affecting things.

**Comment by BurntSushi** [comment]
> having nice names is still achievable

Yes, as I said: 

> I don't think there is a mitigation here. It feels like a "nice to have" aspect of using an enum that we probably wouldn't go out of our way to re-create via other means.

I don't know what the stabilization story is for pattern types or generic integers, but I would prefer we focus on finding a path to stabilization that doesn't require being blocked on hypothetical language features. Or at least, focusing discussion on whether we can utilize them later in a backwards compatible fashion.

I still don't really know what pattern types are (I just don't have time right now to grok it), but generic integer types seem like an implementation detail that's consistent with defining `ascii::Char` as an opaque type?

**Comment by clarfonthey** [comment]
Right, the point here is that any opaque type would be equivalent to the primary hypothetical future implementations. So, explicitly not blocking on any particular path forward, but making sure that they're compatible.

**Comment by scottmcm** [comment]
> I mean, having nice names is still achievable with constants, and this still would make match expressions work as you'd expect.

Though you can `use` a variant, but can't `use` an associated constant.  Thus `use ascii::Char::*;` works with an enum, but not with constants for the names.

I agree that this could absolutely be done with the magic attributes.  But is there anything in particular that you think is *better* with that approach?

The enum version still gets the niche, notably:
```text
[src/main.rs:3] std::alloc::Layout::new::<Option<Option<Option<Option<Option<std::ascii::Char>>>>>>() = Layout {
    size: 1,
    align: 1 (1 << 0),
}
```

**Comment by clarfonthey** [comment]
I will admit that the lack of an ability to `use` associated constants is an immediate downgrade, although I do think that'll eventually be allowed to eventually deprecate the `std::fN::consts` modules.

As far as benefits -- the main ones I see are the ability to unify this type with any one of the potential future features I mentioned. If we stabilise an enum now, we are explicitly locked out of this.

That said, there *is* precedent for separating out the `char` type from the other integer types, and I guess that `ascii::Char` fits right in with that. I'm just not convinced that locking in this as an enum is worth losing the benefits we could get later, especially considering how it could be converted from an opaque type into an enum publicly later if we eventually decide that is better than the alternatives.

**Comment by scottmcm** [comment]
I think it's important that this stay a separate type.  To me, it's an advantage that this *not* just be a `u7`, but that it have the semantic intent behind it of actually *being* ASCII.

For example, if we have pattern types, I don't _want_ to `impl Display for &[u8 is ..128]`, but I think it's possible that we'd add `impl Display for &[ascii::Char]`.  (Like how `Debug` for `char` is very different from `Debug` for a hypothetical `u32 is ..=1114111`.)

Should it be something else to be able to stabilize a subset sooner?  Maybe; I don't know.  I think I'll leave questions like that more to libs-api.

**Comment by clarfonthey** [comment]
That's fair, and in that case I'll concede that the enum works in this case. I still am not sure if it's the best-performing version (I use the `bounded-integer` crate for all sorts of things, and it uses enums as it's codegen) but I feel convinced enough that we can stabilise this as a separate type.

**Comment by safinaskar** [comment]
I don't like name `Char`. It is too similar to `char`, despite these are totally different things. I propose `ASCII` instead

**Comment by safinaskar** [comment]
If you don't like name `ASCII`, then name it `Septet`, but, please, not `Char`

**Comment by BurntSushi** [comment]
I don't think `ASCII` or `Septet` are going to work. I can't imagine those being the names we end up with personally.

`ascii::Char` and `char` are *not* totally different things. They both represent *abstract characters.* One is just a more limited definition than the other, but both are in fact quite limited! `ascii::Char` is namespaced under the `ascii` module, which provides pretty adequate context for what kind of character it represents IMO.

**Comment by clarfonthey** [comment]
Honestly, the name `Ascii` for a type would make sense since it lives in the `ascii` module, since that would mirror existing other Rust types. It doesn't seem out of place to refer to an ASCII character as "an `Ascii`" especially considering how there are already lots of terms that use "ASCII" as a placeholder for "text," as in stuff like "ASCII art." Nowadays, people even refer to Unicode art as ASCII art.

In terms of precedent *for* the `ascii::Char` name, I would point toward the various `Result` types in modules offered across libstd and the ecosystem, where `io::Result` and `fmt::Result` are two common examples. While the base `Result` is in prelude, these separate `Result` types are in separate modules and often used exclusively as module-specific imports, although in rare cases code will directly import them. I don't think this is a good analogy, however, since `ascii::Char` and `char` actually are completely different types with entirely different properties, and not just shorthands for each other, even though you can convert one into the other.

The only real downside I can see to using the name `Ascii` is that it's converting an acronym into a a word and lowercasing most of the letters, which… didn't stop us for stuff like `Arc` and `Rc` and `OsStr`.

Personally, I've thought that the `core::ascii` module has been mostly useless until the addition of this character now that `AsciiExt` is deprecated, although revitalising it to introduce a type *called* `Ascii` would make the most sense to me. Plus, it would even lead a path forward to potentially adding this type to prelude if it were considered useful there, or maybe the preludes of other crates that rely on it.

**Comment by clarfonthey** [comment]
I mean, it does feel weird, but then again, so does calling a vector a "vec" until you say it enough times. If you think of "ASCII" as shorthand for "ASCII char", I don't think that's too far-fetched.

**Comment by kupiakos** [comment]
> 2\. Given that we're overall not fond of `as` casts, and it also allows weirder things like `as i64`, I actually wish the cast could be disabled

I see no issue with this; `ascii::Char`, or whatever it's called, can be fully represented by every numeric type except `bool`. The issue with `as` casts changing the value is irrelevant for this case.

> I mean, it does feel weird

This message is composed of 420 ASCIIs.

**Comment by bluebear94** [comment]
I also support calling this type `Ascii` because calling it `Char` is likely to give worse diagnostics if you mix up `char` and `Char`. The `Char` name would also result in confusion to readers unless you qualify it as `ascii::Char`, which is more verbose than `Ascii`.

Also, rustdoc currently doesn’t play well with types that are exported under a different name from their declaration; for instance, [this method](https://doc.rust-lang.org/nightly/std/ascii/enum.Char.html#method.from_u8) shows the return type as `AsciiChar`, which is the name used in the declaration of this type.

(Note: I dislike the convention of having multiple types named `Error` and `Result` in the standard library and would rather have had `FmtError`, `IoError`, and so on, so my opinions might be biased.)

**Comment by clarfonthey** [comment]
I'm actually kind of fine with `io::Result` and `io::Error` since you tend to have lots of different error and result types, whereas something as fundamental as `ascii::Char` feels more like something you'd want to avoid qualifying on import.

Hence why I think that exporting it as just `Ascii` is best, since `AsciiChar` feels "redundant" in some sense. Even though "Ascii" by itself could refer to a string, since the string version is so easy to create (`[Ascii]`), it feels okay.

**Comment by joshtriplett** [comment]
FWIW, I really don't think we should expose these character names as enum variants to the user. We should encourage just using character constants; if we need a way to write this type directly (rather than `b'.'` which produces a `u8`), we should give a special character syntax for that (e.g. `a'.'`) rather than named enum variants.

**Comment by joshtriplett** [comment]
> Using `Ascii` as a singular noun sounds very strange to me. "ASCII art" is using "ASCII" as an adjective. I'm not aware of any wide usage of "Ascii" as a singular noun referring to a single character. Its usage as a noun, in my experience, refers to a character set and/or encoding. Not a member of that set/encoding.

Agreed. calling it `AsciiChar` seems preferable, to give clear error messages and give a type that makes sense unqualified.

**Comment by BurntSushi** [comment]
> > Using `Ascii` as a singular noun sounds very strange to me. "ASCII art" is using "ASCII" as an adjective. I'm not aware of any wide usage of "Ascii" as a singular noun referring to a single character. Its usage as a noun, in my experience, refers to a character set and/or encoding. Not a member of that set/encoding.
> 
> Agreed. calling it `AsciiChar` seems preferable, to give clear error messages and give a type that makes sense unqualified.

I still personally like `ascii::Char`. It avoids the redundancy in naming. I concede that its unqualified use may be confusing, but I think the antidote to that is to use its qualified form.

**Comment by kupiakos** [comment]
> FWIW, I really don't think we should expose these character names as enum variants to the user. We should encourage just using character constants; if we need a way to write this type directly (rather than `b'.'` which produces a `u8`), we should consider giving a special character syntax for that (e.g. `a'.'`) rather than named enum variants.

I agree for eventual stabilization that a dedicated `a'.'` for `AsciiChar` and `a"."` for `&[AsciiChar; N]`  would be best, but for now, in nightly, these variants should be exposed until such a syntax is accepted and merged. Otherwise, there's no infallible way to construct an `AsciiChar` and no easy way to `match` on one; you have to deal with optimized-away `AsciiChar::from_u8(b'A').unwrap()` or match on `u8` instead of `AsciiChar`.

**Comment by kupiakos** [comment]
> Agreed. calling it `AsciiChar` seems preferable, to give clear error messages and give a type that makes sense unqualified.

I'm pro-`AsciiChar` though not anti-`ascii::Char`. A name that's easier to understand unqualified seems less risky for the stdlib.

**Comment by kupiakos** [comment]
> ```rust
>     const fn as_u8(self) -> u8;
>     const fn as_char(self) -> char;
> ```

Hmm, shouldn't these be named `to_`? The [API guidelines](https://rust-lang.github.io/api-guidelines/naming.html) describe `as` as used for borrowed -> borrowed names, and `to_` for owned -> owned Copy types. Also see:

- [`char::to_digit`](https://doc.rust-lang.org/std/primitive.char.html#method.to_digit)
- [`u32::to_be`](https://doc.rust-lang.org/std/primitive.u32.html#method.to_be)

**Comment by joshtriplett** [comment]
*Zero* objection to these variants being exposed in nightly, just to the idea of stabilizing them.

**Comment by scottmcm** [comment]
> I see no issue with this; `ascii::Char`, or whatever it's called, can be fully represented by every numeric type except `bool`.

This is an interesting point.  Using `x as ffi::c_char` would work properly no matter whether `char` is signed or unsigned on the target, for example.  (Well, barring ebcdic, I guess, if you care about such things.)

**Comment by ChrisDenton** [comment]
One thing that occurs to me that it might be useful to have [encode_utf16](https://doc.rust-lang.org/std/primitive.str.html#method.encode_utf16) implemented for `[AsciiChar]`. Because that's simply `u8 -> u16` for each char.

**Comment by clarfonthey** [comment]
Isn't `c.as_u8() as u16` good enough for that?

Ideally we'd be able to just say `c as u16` but I'm not sure if that works at the moment.

**Comment by clarfonthey** [comment]
re: #113295, `Step` is now implemented for `ascii::Char` as well.

**Comment by Kimundi** [comment]
About naming: I agree that a short unique unqualified name would probably be best name, due the expected usage and the possibility of having it in the prelude at some point. However, seeing how `Ascii` and `Char` are two extremes with issues, how about we split the difference and call it `Achar` ? Would not be the first type that had the compromise of using a single character prefix to differentiate it: `iN` vs `uN`, C/C++s `char, wchar_t, char8_t, char16_t, char32_t`, etc.

---

About the Implementation that currently is in std: I see that `Debug` is derived on the enum, while `Display` use the string repre. Would it not make sense to have the Debug output be like for `char`?

Consider a `Debug` printout of a `[ascii::Char]` - what is more useful:
- `['H', 'e', 'l', 'l', 'o']`
- `[CapitalH, LowerE, LowerL, LowerL, LowerO]`

**Comment by bluebear94** [comment]
I disagree with calling it `Achar` (though I at least prefer it to `Char`), but I agree with changing the `Debug` impl to print the string repr.

**Comment by rben01** [comment]
If going the `Achar` route, I think it should instead be spelled `AChar` because “ASCII” and “Char” are two separate words. Or even `achar`, all lowercase, as this type feels pretty primitive. That said, I'd prefer `AsciiChar` (or even `Char7` 😱). Not a fan of `Ascii` — it sounds like it refers to an ASCII string, not a single char.

Regarding `Debug` printing, wrapping the result of [`char::escape_default`](https://doc.rust-lang.org/std/primitive.char.html#method.escape_default) in single quotes seems like it'd be optimal (except maybe `"` should be `'"'` instead of `'\"'`), regardless of the underlying implementation of `ascii::Char`.

One question I have is if it would be possible to match against ranges of `ascii::Char`. I could see something like `match c { Char::CapitalA..=Char::CapitalZ => ..., /* other cases here */ }` being useful, but currently range patterns only support `char` and numeric types. Without this, I could see people falling back to `match c.as_u8() { b'A'..=b'Z' => ..., /* other cases here */, 128..=255 => unreachable!() }` to satisfy the compiler.

**Comment by AngelicosPhosphoros** [comment]
>EDIT: Is there a reason that matching against any range pattern a..=b of a PartialOrd type isn't just equivalent to x if (a..=b).contains(&x)?

Yes, range based match allows to check matches for totality.

**Comment by clarfonthey** [comment]
You can't match ranges of *any* custom-defined type at the moment because, while we do have a mechanism for doing this (see: #74446) with value matches (deriving `Eq` will let you do this), it does not work for range matches (there's no equivalent for deriving `Ord`).

I *do* think that it would be nice to have ranges work for this, though, especially given how the values are effectively integers. It's quite unfortunate that you have to call `as_u8` in order to make ranges work, since it means that you always need to add an unreachable branch for `128..=255`.

**Comment by Finomnis** [comment]
Missing for usage as generic string parameter:
- `impl AsRef<str> for [ascii::Char]` 
- `impl<const N: usize> AsRef<str> for [ascii::Char; N]`

**Comment by llogiq** [comment]
Also missing:

* `impl core::str.:pattern::Pattern for ascii::Char`
* `impl core::str.:pattern::Pattern for &[ascii::Char]` (And perhaps `Vec<ascii::Char>` and `Box<[ascii::Char]>`?)
* optional: `impl<const N: usize> core::str.:pattern::Pattern for [ascii::Char; N]` for inline storage of data, not sure if it's worth it though

**Comment by Sky9x** [comment]
half a pre-RFC on ascii literals: https://internals.rust-lang.org/t/pre-rfc-ascii-literals/19960

**Comment by clarfonthey** [comment]
> re: #113295, `Step` is now implemented for `ascii::Char` as well.

Just want to point this out again since the original description hasn't been edited.

**Comment by mina86** [comment]
How about `from_digit` method, see https://github.com/rust-lang/rust/pull/118963.

**Comment by kupiakos** [comment]
> How about `from_digit` method, see #118963. Though admittedly, @Sky9x’s proposal of `a` prefix would alleviate need for it since writing `a"0123456789abcdef"[digit]` isn’t that bad.

There's currently a [`ascii::Char::digit`](https://doc.rust-lang.org/std/ascii/enum.Char.html#method.digit) method, but its API doesn't match [`char::from_digit`](https://doc.rust-lang.org/std/primitive.char.html#method.from_digit) with its `radix` parameter. `ascii::Char` also has a `digit_unchecked` which doesn't have an equivalent in `char`. 

Unless there's any opposition, I propose that:

- `ascii::Char::digit` be renamed and rewritten to match `char::from_digit`, using the same implementation.
- `ascii::Char::digit_unchecked` is removed; users can instead write `from_digit(x, 10).unwrap_unchecked()`.

**Comment by mina86** [comment]
> And so long as the compiler is showing that as just `<u8 as From<Char>>`, that's a bit awkward to read, and might be evidence that it would be nice for the type name, not just the namespace, to be clearer here.

Note though that error messages can be changed while APIs cannot. Long-term, a better API is more important than better diagnostics. (I have no strong opinions on naming though I lean towards `ascii::Char` since I dislike when symbols include module they belong to in their names).

**Comment by clarfonthey** [comment]
My main motivation for naming it `Ascii` instead of `Char` or `AsciiChar` is because it doesn't need the module to describe what it is, but also isn't overly long. Most people will be importing it to use without a module, and it should describe what it is fully in these cases.

Generally, the only cases where types use ambiguous names are when tey're specific to a module and not useful outside it. For example, `io::Result ` and `io::Error` are not general-purpose types for any program, but only for I/O. There's also the case of types that are rarely named, like `array::IntoIter` and `slice::Iter`, which are almost always used implicitly as the result of calling methods.

In this case, `Char` doesn't fit this, really. ASCII text is a fundamental data type meant to be used outside the `ascii` module. Many people will be directly naming this type in their code and shouldn't have to distinguish `char` and `Char`, or use the module as `ascii::Char`.

Maybe another example is types which have the same name but aren't meant to be used together, e.g. types in `sync` and `async` modules. Here, you generally want exactly one module's types, and code that uses both will often be a library trying to provide compatibility rather than any one program.

So, it's this reason I support `Ascii`, `AChar`, or `AsciiChar`, in that order.

Maybe even a better example is `BTreeMap` and `HashMap`; we could make these `btree::Map` and `hash::Map` but we don't because perfectly reasonable programs might use both.

**Comment by sffc** [comment]
I don't see it mentioned yet in this thread so I'll just say that we use `AsciiByte` in the icu4x code base. It is a range-limited byte (`u8`), not a range-limited `char`.

On this line of reasoning, `AsciiU8` could be another choice, similar to `NonZeroU8`.

**Comment by robertbastian** [comment]
`U7`

**Comment by mina86** [comment]
I agree, that would be ideal.

What we need is support for casting integer literals to custom types.  Good example of what I mean is:

```rust
// Compiles:
let num: core::num::NonZeroU8 = 42;
// Doesn’t compile:
// let num: core::num::NonZeroU8 = 0;
```

This then would need to be extended to handle byte literals as well such as the following would work as well:

```rust
let num: core::num::NonZeroU8 = b'*';
```

With all that, it would be trivial to define `U7` which would work as expected without a need to introduce any new APIs:

```rust
let chr: core::num::U7 = b'*';
let text: &[core::num::U7] = core::slice::from_ref(&chr);
let string: &str = text.into();
```

The question is: what’s a plausible timeline for such a feature (if it’s even plausible to be implemented) compared to how badly we want an ASCII character type.

**Comment by chloekek** [comment]
The name U7 suggests a general purpose 7-bit integer type and would therefore imply a Display impl that formats the number in decimal, as opposed to an ASCII character.

**Comment by mina86** [comment]
> as it doesn't adhere to Rust’s principle of making most everything explicit.

Rust is happy to not adhere to that principle (I really question whether it even is a principle).  Heck, integer literals are example of implicit type coercion.  It’s really not an issue.

> For instance, what do I hover over in my IDE to see the method that is used to turn 42 into a NonZeroU8?

One option is hovering over the literal.

**Comment by programmerjake** [comment]
> The name U7 suggests a general purpose 7-bit integer type and would therefore imply a Display impl that formats the number in decimal, as opposed to an ASCII character.

plus, I'm hoping rust eventually gets [generic integer types](https://github.com/rust-lang/rfcs/pull/2581), so `u7` (or at least `uint<7>`) would be a real 7-bit integer type that behaves like an integer, not a character. We'd still want a character type for ASCII otherwise we could just use the pattern type `u8 is ..0x80` once [pattern types](https://github.com/rust-lang/types-team/issues/126) are added.

**Comment by sffc** [comment]
> While that interpretation makes sense, I'm not a fan. Putting `U8` in the name implies that it's a predominantly numeric type. Rather, it is a displayable character limited to the ASCII range; it being `u8`-sized is solely a consequence of it being limited to ASCII. That's a property that makes sense to put front-and-center for zerocopy de/serialization, but IMO is more confusing than helpful in the stdlib, since the goal is to treat it as a _character_ and not an _unsigned number_. ... It follows many of the same restrictions as `char`. It's displayed like a `char`. A sequence of these forms a `str`. Keeping the "char" name makes tons of sense to me.

This line of reasoning makes sense to me, though it's worth pointing out that slices of this new type are valid as _both_ `&[u8]` _and_ `&str`.

About 1/4 of the ASCII range is not displayable. A valid interpretation of the type is that it utilizes more human-readable rendering _when possible_, but it is still just a byte.

One aspect in favor of `char` is that `char` is very nichey and so is the Ascii Char.

> We shouldn't `impl Add<Rhs = Self>` or other arithmetic on `Self` for this type for the same reasons as `char`.

Doing arithmetic on these is useful (case conversion, etc), but it is easy enough to convert back and forth to `u8` for this. I agree it may be more error-prone to allow these by default on this type.

**Comment by mina86** [comment]
> We shouldn't impl Add or other arithmetic on Self for this type for the same reasons as char. It follows many of the same restrictions as char.

I disagree. As I expressed in [my comment in PR implementing Add](https://github.com/rust-lang/rust/pull/120219#issuecomment-1908338791), ASCII characters have no holes in the domain (char forbids surrogates) and all 7-bit numbers are valid ASCII characters (char doesn’t fit in 20 bits but doesn’t cover all 21 bits).

> Doing arithmetic on these is useful (case conversion, etc), but it is easy enough to convert back and forth to u8 for this. I agree it may be more error-prone to allow these by default on this type.

It’s easy to convert to `u8`.  It’s hard to convert from `u8`.  This would be true if Rust had `core::num::U7`.

**Comment by rben01** [comment]
Arithmetic should not exist on `AsciiChar` (or whatever it ends up being called). Any arithmetic common and useful enough to be worth thinking about, like addition with `'a' - 'A'` for case conversion, should just be exposed through a method like `to_ascii_uppercase` (or just `to_uppercase` since the domain is already limited to ASCII).

> I disagree. As I expressed in https://github.com/rust-lang/rust/pull/120219#issuecomment-1908338791, ASCII characters have no holes in the domain (char forbids surrogates) and all 7-bit numbers are valid ASCII characters (char doesn’t fit in 20 bits but doesn’t cover all 21 bits).
>
> There is a difference between char, NonZero and AsciiChar in regards to Add. AsciiChar can be thought of as u7 where addition has an obvious wrapping behaviour of keeping only the seven least significant bits of the result. There is no such clear rule for char or NonZero. char is not quite 21-bit and both have forbidden representations less than their max value.

I certainly would not expect wrapping. If adding 1 to char 127 got me back to `'\0'` I'd be very very surprised. If anything I'd expect a panic, as with all other default arithmetic operations. (Now, should there be `Wrapping<AsciiChar>`? I'd still say no, but at least its behavior wouldn't be surprising.)

**Comment by mina86** [comment]
> I certainly would not expect wrapping. If adding 1 to char 127 got me back to '\0' I'd be very very surprised. If anything I'd expect a panic, as with all other default arithmetic operations.

And you would get behaviour as with any default arithmetic operations.  `AsciiChar(127u8) + 1` would produce the same effect as `AsciiChar(255u8 + 1)`.  Why wrapping in AsciiChar would surprising where wrapping with u8 isn’t?

**Comment by Finomnis** [comment]
> > I certainly would not expect wrapping. If adding 1 to char 127 got me back to '\0' I'd be very very surprised. If anything I'd expect a panic, as with all other default arithmetic operations.
> 
> And you would get behaviour as with any default arithmetic operations. `AsciiChar(127u8) + 1` would produce the same effect as `AsciiChar(255u8 + 1)`. Why wrapping in AsciiChar would surprising where wrapping with u8 isn’t?

`u8` does not wrap, it panics on wrap (on debug builds)

**Comment by sffc** [comment]
If `AsciiChar(127u8) + 1` panics in debug, like other overflowing integer types do, what does it do on release? It can't just roll over to `128` because that would be UB since 128 is not a valid `AsciiChar`, and rolling to `\0` is not free (still requires branching, unlike unsigned int overflow which is handled in hardware). So it seems cleaner to not `impl Add for AsciiChar`.

**Comment by clarfonthey** [comment]
What exactly is the intent behind some kind of `Add` operation that isn't covered by `Step` and a `to_digit` method?

**Comment by mina86** [comment]
> If AsciiChar(127u8) + 1 panics in debug, like other overflowing integer types do, what does it do on release? It can't just roll over to 128 because that would be UB since 128 is not a valid AsciiChar, and rolling to \0 is not free (still requires branching, unlike unsigned int overflow which is handled in hardware).

It wraps to zero which doesn’t require branching.  It’s just an and operation.  Do you also argue that Index shouldn’t do bound checking because those introduce branch?

> What exactly is the intent behind some kind of Add operation that isn't covered by Step and a to_digit method?

This is disingenuous question. You might just as well ask what is intend behind some kind of AsciiChar which isn’t covered by char::is_ascii, str::is_ascii and str::from_utf8? If you really think that Step, a trait designed for iterators, is appropriate for use with arithmetic than I really don’t know what to tell you.

**Comment by kupiakos** [comment]
> About 1/4 of the ASCII range is not displayable. A valid interpretation of the type is that it utilizes more human-readable rendering _when possible_, but it is still just a byte.

This doesn't seem particularly relevant - this is just a difference between Unicode and ASCII. U+202E, anyone?

> what does it do on release

Overflow from 128 -> 0, with the associated code cost to do that.

> still requires branching, unlike unsigned int overflow which is handled in hardware

The overflow is branchless: it's `transmute((x as u8 + 128 + y as u8) & 0x7f)`. It's more costly than a `u8` add, but it's not terrible.

> I disagree. As I expressed in [my comment in PR implementing Add](https://github.com/rust-lang/rust/pull/120219#issuecomment-1908338791), ASCII characters have no holes in the domain (char forbids surrogates) and all 7-bit numbers are valid ASCII characters (char doesn’t fit in 20 bits but doesn’t cover all 21 bits).

This makes a solid case for the differences with `char`/`NonZero*`, though I'm still skeptical it's the right choice for Rust long-term. It's the only numeric add where release-profile overflow isn't just handled by the hardware. I suspect there are common cases that would be lighter-weight to do the arithmetic as `u8`, and then convert to `AsciiChar`, then to perform multiple `AsciiChar` wrapping operations. (Relatedly, I'm broadly against adding custom-bit-width integers to Rust. Eager masking/wraparound is, IMO, a niche need that libraries should provide for.)

I think we should _discourage_ users from writing code like `x + a'0'` or `c + (a'a' - a'A')` and push them towards helper methods that can be more trivially optimized.

> It’s easy to convert to `u8`. It’s hard to convert from `u8`.

I challenge "hard". It's fallible, sure, but we could provide a `fn from_u7(x: u8) { unsafe { transmute(x & 0x7f) } }` for a safe, infallible "I'm assuming that top bit is unset, just truncate to convert".

**Comment by kupiakos** [comment]
> This is disingenuous question. You might just as well ask what is intend behind some kind of AsciiChar which isn’t covered by char::is_ascii, str::is_ascii and str::from_utf8

It's a valid question. The `Step` implementation is O(1), it's just less ergonomic than `+ usize`. The intent and real-world use cases for `AsciiChar` have been [addressed and described](https://doc.rust-lang.org/std/ascii/enum.Char.html#when-to-use-this). Implementing controversial features requires answers to "Why should we have this when there are workable alternatives?"

From what I see the primary advantage that `impl Add for AsciiChar` provides is developer ergonomics. I'm skeptical that the overflow behavior itself is something that anyone would actually take advantage of instead of it being incidental (and more costly) behavior. It's also not clearly obvious to users. But perhaps the ergonomics are worth it!

I'm sure we'd all appreciate a set of real-world use cases for `impl Add for AsciiChar` that _wouldn't_ be better suited as a helper function or a conversion to/from `u8`. We don't currently have an equivalent to `+ b'a'`/`+ b'A'` in `char`, so maybe that's one of them? A before/after comparison could directly compare the ergonomics.

**Comment by mina86** [comment]
> It's a valid question. The Step implementation is O(1), it's just less ergonomic than + usize. The intent and real-world use cases for AsciiChar have been [addressed and described](https://doc.rust-lang.org/std/ascii/enum.Char.html#when-to-use-this). Implementing controversial features requires answers to "Why should we have this when there are workable alternatives?"

It’s not because it presupposes that if something can be done with Step or to_digit than the alternatives are to be dismissed.  Getting nth letter and Ascii85 implementation would benefit from Add but of course it all can be done with `AsciiChar::new(...).unwrap()`.  If you don’t think ergonomics are better/worth it than there’s nothing I can say to convince you otherwise.

**Comment by clarfonthey** [comment]
> > It's a valid question. The Step implementation is O(1), it's just less ergonomic than + usize. The intent and real-world use cases for AsciiChar have been [addressed and described](https://doc.rust-lang.org/std/ascii/enum.Char.html#when-to-use-this). Implementing controversial features requires answers to "Why should we have this when there are workable alternatives?"
> 
> It’s not because it presupposes that if something can be done with Step or to_digit than the alternatives are to be dismissed. Getting nth letter and Ascii85 implementation would benefit from Add but of course it all can be done with `AsciiChar::new(...).unwrap()`. If you don’t think ergonomics are better/worth it than there’s nothing I can say to convince you otherwise.

The point here is that `Add` fundamentally is a nonsensical operation for *ASCII characters*, even though the type internally is just a byte whose value can be added. `char` doesn't support `Add` for similar reasons.

The reason why I mentioned `Step` is because things like getting the nth letter *are* covered by it; `'a'..='z'.nth(x)` works for `char` just fine and it would also work just as well for `AsciiChar` too. I would even argue that this form is clearer than using addition, and less error-prone.

I don't see how implementing addition gives you any positive ergonomics, and there's even the additional confusion of what you're adding: `u8`? other ASCII characters? It just introduces so many problems that feel like they could be easily solved by other methods, which is why I'm asking what problems you'd like to solve that aren't covered by those methods.

**Comment by joseluis** [comment]
I just want to point out that `AsciiChar` should derive `Default` returning the `Null` variant, matching `char`'s implementation returning `'\x00'`, since I've not seen it mentioned anywhere and right now it lacks it.

**Comment by clarfonthey** [comment]
I also hate to be a broken record but `Step` is still not mentioned anywhere in the issue description, and it should be.

**Comment by programmerjake** [comment]
i think we should add `[Char; N]::as_bytes()` which returns `&[u8; N]`

**Comment by programmerjake** [comment]
I would write it `Ascii<u8>` instead of `Ascii<char>` since the latter makes me think it's just a 32-bit value range limited to `0..=0x7F`, similarly for arrays. For unknown-length, idk if we use `[u8]` or `str`...
one benefit of `Ascii` is a owned string type easily falls out: `Ascii<Vec<u8>>`/`Ascii<String>`

**Comment by programmerjake** [comment]
you can't simultaneously `AsRef<str> + AsRef<char>` since they're encoded differently, unless you have duplicated storage (which I think we can agree is not what we want)

**Comment by programmerjake** [comment]
If we have `Ascii<T: ?Sized>: Deref<Target = T>`, I think we'll need both `Ascii<str>` and `Ascii<[u8]>` as well as the corresponding owned types.

**Comment by Kimundi** [comment]
Heh, that actually reminds me of the API I came up with for a toy hex formatting crate of me: https://docs.rs/easy-hex/1.0.0/easy_hex/struct.Hex.html

Basically, the whole api resolves around being able to cast `T` to/from Hex<T> just to get changed trait impl semantic. In my case its just a repr(transparent) wrapper, so I do not change representation, but still the idea seems similar.

That said, I feel like a basic `AsciiChar` type is still the best course of action, otherwise it seems like the whole API discussion here has to be started from scratch :D

**Comment by jongiddy** [comment]
Will it be possible to backwards-compatibly redefine `[u8]::escape_ascii` to return an `Iterator<Item=AsciiChar>` instead of the current `Iterator<Item=u8>`?

Currently the returned iterator provides a `to_string()` method that collects the characters into a string, but if any other iterators are chained, we lose the info that the bytes are safe to collect into a String.

**Comment by programmerjake** [comment]
> Will it be possible to backwards-compatibly redefine `[u8]::escape_ascii` to return an `Iterator<Item=AsciiChar>` instead of the current `Iterator<Item=u8>`?

yes, but likely only if rust implements something like edition-dependent name resolution where the same name resolves to two different `escape_ascii` functions depending on the edition.

I previously proposed something like that: https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/Effect.20of.20fma.20disabled/near/274199236

**Comment by clarfonthey** [comment]
This is a random thought I had, but seeing the progress on this makes me wonder if we could make `char` itself implemented as a lang item struct instead of a primitive in the language without breaking anything.

Figuring out what would have to be done so avoid any breakages there could probably be useful here, since they would also need to be applied to ASCII chars too.

**Comment by clarfonthey** [comment]
Getting back to this feature a bit. Genuine question: instead of the common named variants, now that the precedent has been set for `CStr` having dedicated `c"..."` literals, would it be reasonable to add `a'x'` and `a"..."` literals? This could also solve the issue with `ascii_char_variants` being unstable, preferring that you use the literals instead of the variants.

Not sure if implementing this as a POC in unstable would require an RFC or not.

**Comment by leb-kuchen** [comment]
is it possible to hide the variants in the sidebar of the documentation or even better the entire section?

**Comment by NathanielHardesty** [comment]
Re: ASCII Literals

As an end user of Rust who wants good ASCII support, I believe the best way to improve this feature is to add literals such as `a'A'` and `a"ASCII"` as others have suggested. Doing fancy manipulations on strings is far less important (to me anyways) than just being able to write strings so that they can be stored in structs or passed into functions.

Currently, the best solution to arriving at the ASCII subset is this:

- `[Char::SmallA, Char::SmallS, Char::SmallC, Char::SmallI, Char::SmallI]`

Being forced to write out all of the `Char` variants just to arrive at a `[Char]` is cumbersome and it seems as though these variants are an unstable feature unto themselves which is best avoided.

On the other hand, byte string literals can be used in this manner:

- `b"ascii".as_ascii().unwrap()`
- `unsafe {b"ascii".as_ascii_unchecked()}`

This forces me to use runtime checks, or simply cross my fingers, rather than use compile time checks in order to assure that the ASCII subset is not being violated. I'm perfectly fine with doing runtime checks when I'm parsing what very well could be arbitrary data, but source code is not arbitrary and violations should be able to be caught at compile time.

**Comment by ChayimFriedman2** [comment]
@NathanielHardesty If the functions for constructing `&[AsciiChar]` from `&[u8]` are const, you can create a macro for it (it's still worse than compiler support, but not by much).

**Comment by Marcondiro** [comment]
Hello, do you plan to add a [char::to_digit(self, radix: u32)](https://doc.rust-lang.org/std/primitive.char.html#method.to_digit) equivalent?
This method was proposed in https://github.com/rust-lang/rust/issues/95969 / https://github.com/rust-lang/rust/pull/96110 for `u8` but was rejected.
Thanks!

**Comment by NobodyXu** [comment]
@Marcondiro you will need to open an issue in [rust-lang/libs-team](https://github.com/rust-lang/libs-team/), which is called an ACP.

The libs-api team would then provide you with feedback there.

**Comment by zopsicle** [comment]
@CodesInChaos `&[NonZero<ascii::Char>]` cannot be converted to `&CStr` because there is no terminating nul. It can be infallibly converted to `CString` though.

**Comment by clarfonthey** [comment]
`NonZero<ascii::Char>` actually could be implemented with just a library change, although I'm not sure whether libs-api would rather make that go through ACP first. I would imagine that, at least this being the first non-primitive type to support `NonZero`, it would be nice to ensure that this precedent is okay.

**Comment by nnethercote** [comment]
A drive-by bikeshed comment that I don't think has been made before: `UpperCaseA` and `LowerCaseA` would be preferable to `CapitalA` and `SmallA`, for consistency with `char::is_{upper,lower}_case`, and because they are just clearer names.

**Comment by tmke8** [comment]
Would it make sense to also add

```rust
impl String {
    pub fn push_ascii(&mut self, ch: ascii::Char) {
        // SAFETY: an ASCII char is valid UTF-8
        unsafe { self.as_mut_vec().push(ch.to_u8()) };
    }
}
```

?

**Comment by LunarLambda** [comment]
i think it would be fine to just do `string.push(ascii.to_char())` or whatever the conversion method is. The optimizer should be able to remove the `if char < 128` check so it should be cheap.

**Comment by CodesInChaos** [comment]
I think `Extend<AsciiChar>` should be implemented for string. Then `extend_one` would fill the role of your `push_ascii`.

**Comment by Kixunil** [comment]
I was looking into this wondering why it's unstable. I'd really love to see at least some minimal stabilization. IIUC the main blockers are:

* Should it be `enum`?
* `Char` can be confused with `char`, what's the solution?
* Should conversion methods return `Option` or `Result`?

My thoughts: the sentiment is generally against `enum` claiming that we can do anything with opaque struct that we can do with `enum`. But this claim is invalid. While some people consider `as` undesirable, even if it was, it still makes exhaustive `match`. Unless the type becomes language type it needs to be an `enum` to allow exhaustive matching. So I personally prefer `enum` but I prefer stabilizing sooner even more. Changing `struct` to `enum` is currently MSRV breaking but that can be fixed just by allowing `Foo { .. }` patterns to match `enum`s.

What wasn't mentioned is the ability to rename types using `use ... as ...`. There already are "conflicting" names in the library already mentioned in the discussion, however an interesting point is that e.g. `serde` imports `std::error::Error` as `StdError`. I think that's very sensible approach. So one solution is to recommend renaming for people who don't want to use `ascii::Char`. I actually co-maintain a library where we have `absolute::Time` and `relative::Time`. We always use the module names in the doc and recommend people do so in their code, though anyone is free to import the name or just rename the import to `AbsoluteTime` or whatever they like. It works well so far. `core` could do the same. Just use the full name in examples and recommend that people do this. Also the compiler messages when `char` mismatches `Char` could be special cased to call out the one-letter difference.

Error type being needed anyway for `TryFrom` is a good argument that the conversions might as well return it in `Result` instead of `Option`. It makes `?` conversions somewhat easier. Though, if the internal type is not going to carry the input then `ok_or` is a bit more convenient way to add it back than `map_err(|_| ...)`.

**Comment by oxalica** [comment]
> * `Char` can be confused with `char`, what's the solution?

Would it be better to just call it `std::ascii::Ascii` so users can `use` it without confusion? We would likely NOT to have a dedicate ASCII-str because `[ascii::Char]` suffices. To me it is redundant for writing `ascii::Char` where there is no `ascii::Str` or `ascii::OtherEncoding` to disambiguate with. All our API names `as_ascii`, `to_ascii`, `from_ascii` are already hinting that they should return "Ascii" (or `[Ascii]`?).

> What wasn't mentioned is the ability to rename types using `use ... as ...`.

Not sure if it's only me, but it's too hard to come up a new name, unless I have to do so in public interface. 😅 I always use a prefix mod to disambiguate `use std::io; io::Error`, or just give up and write full name `std::error::Error`.

**Comment by tmke8** [comment]
>the sentiment is generally against enum claiming that we can do anything with opaque struct that we can do with enum. But this claim is invalid. While some people consider as undesirable, even if it was, it still makes exhaustive match. Unless the type becomes language type it needs to be an enum to allow exhaustive matching.

Is it possible to have it initially be an enum (in order to get exhaustiveness checking for free) but to keep the internals hidden so that it can be made a language type later? (This would require constants for all the enum variants because they would be hidden.)

I also wanted to mention that in the situations where I want to use this, I actually only care about the printable ASCII characters and would prefer the control characters weren't included. Not sure what other people think here.

**Comment by clarfonthey** [comment]
While it would be cool to have a version of this type that omits control characters, I don't think that this is a use case the standard library should support, at least until/if some form of pattern types is stable.

The primary goal of this type, as I understand it, is to be able to remove the UTF-8 validation penalty when converting slices of ASCII characters to strings. If you know for a fact that your slice of bytes is ASCII-only, then you can convert it to a `&str` without any extra validation while still retaining some of the nice properties of ASCII strings, like being able to slice them at any index without disrupting UTF-8 validity.

**Comment by Kixunil** [comment]
@oxalica as I understand it, "ASCII" is an adjective so `ascii::Ascii` sounds weird. But if this is going to make the difference between stabilizing in two years and stabilizing after two years the I'd be able to tolerate it.


> I always use a prefix mod to disambiguate `use std::io; io::Error`, or just give up and write full name `std::error::Error`.

Same here and I think it's fine, even elegant. I see no problem why `ascii::Char` would be bad.

> Is it possible to have it initially be an enum (in order to get exhaustiveness checking for free) but to keep the internals hidden so that it can be made a language type later? (This would require constants for all the enum variants because they would be hidden.)

I don't think so, you need the variants exposed to check exhaustiveness. But why would it need to be a language type? For literals it should only need to be a lang item.

@clarfonthey exactly!

**Comment by lolbinarycat** [comment]
I think it would make sense to put `#[doc(alias("AsciiChar"))]` on this, since that's the reexported name its used under in other parts of the standard library.

**Comment by ChaiTRex** [comment]
@Finomnis [said](https://github.com/rust-lang/rust/issues/110998#issuecomment-1826304417):
> Missing for usage as generic string parameter:
> * &vellip;
> * `impl AsRef<str> for Vec<ascii::Char>` (not sure about this one, it's only useful if passing ownership. And then there's no point in keeping it a `Vec` as it could as well be converted to a `Box`)

I've used `impl AsRef<str> for String` in my own code when I didn't need the `String` afterward. `Vec` is likely to be heavily used with ASCII data, and converting to a `Box` has a cost, as [it calls `shrink_to_fit`](https://doc.rust-lang.org/src/alloc/vec/mod.rs.html#1496-1504), possibly reallocating, so it shouldn't be a necessity to convert to a `Box` to gain any features.

@CodesInChaos [said](https://github.com/rust-lang/rust/issues/110998#issuecomment-2483646269):
> 1. Why do the fallible functions return `Option` instead of `Result<T, AsciiError>`? Returning `Result` would be consistent with utf8 conversion, and would produce a reasonable error when unwrapping. Plus `TryFrom` already needs an error type.

It's a common thing to return `Option` when there's only one cause of failure (here, the `u8` was not ASCII) and there's no further information to provide, like with [`char::from_u32`](https://doc.rust-lang.org/std/primitive.char.html#method.from_u32). As an example where `Result` is used, [`String::from_utf8`](https://doc.rust-lang.org/std/string/struct.String.html#method.from_utf8) provides [information about where in the bytes the UTF-8 goes bad](https://doc.rust-lang.org/std/string/struct.FromUtf8Error.html), so if we were to do `Vec::<ascii::Char>::from_ascii`, a `Result` would be good to match existing behavior.

**Comment by ChaiTRex** [comment]
If this is going to be stabilized, I strongly agree with adding the `a'x'`, `a"xyz"`, and `ar#"xyz"#` notations, as:

* ASCII is frequently used, so making it easy to use is a good thing.
* I'm currently adding `u8`'s `is_ascii*` methods to `ascii::Char` and, for the doctests, I have to decide between spending extra time looking up the name of each enum variant in a long list or using the bulky `ascii::Char::from_u8(b'A').unwrap()` a bunch of times.

**Comment by clarfonthey** [comment]
I feel like that performance issue, at least for now, could be fixed with some inlining and the `is_statically_known` intrinsic. At least, I don't think that people should be forced to change the way they write this if we can just have the compiler fix the performance for us.

**Comment by programmerjake** [comment]
> I was looking at this, and now I wonder if a potential `a'c'` and `a"Ascii string"` for `Char` and `&'static [Char]` respectively would be a good idea. It might be worth me opening another ACP for that?

Those are new language features so IIRC they'd require a RFC.

**Comment by hanna-kruppe** [comment]
> I was looking at this, and now I wonder if a potential `a'c'` and `a"Ascii string"` for `Char` and `&'static [Char]` respectively would be a good idea. It might be worth me opening another ACP for that?

It seems premature to design a new literal kind for a library type that is still unstable and not (yet) widely used. C string literals came many years after `&CStr` . There's no rush to have literals at the same time as the type -- checked conversion from `&str` or `&[u8]` to `&[ascii::Char]` can be done at compile time with `const fn`, so literals would only saving a little bit of typing. Very little typing, if you're willing to define a macro that makes `b!("hello")` and `b!('a')` work.

**Comment by kupiakos** [comment]
> > Hmm, shouldn't these be named to_? The [API guidelines](https://rust-lang.github.io/api-guidelines/naming.html) describe as as used for borrowed -> borrowed operations, and to_ for owned -> owned operations on Copy types.
> 
> The api guidelines also say that `as_` should be used for "free" conversions, which this definitely is. See `as_fd` and `ThreadId::as_u64`.

The [API naming guidelines](https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv) also say that `as_` should be used for borrowed -> borrowed conversions, which this _definitely is not_.

`as_fd` is generally a method on non-`Copy` types, and is conceptually borrowing the input - i.e. the fd's validity depends on `&self` remaining valid. `ThreadId::as_u64` is nightly and so its name carries no weight at the moment. `to_u64` may be the more proper name for that function, given that `ThreadId` is `Copy`.

Consider `f32::to_bits` or `i32::to_ne_bytes`: they are both "free" conversions, and yet they use `to_`, since the input is a `self` type that is `Copy`.

I do think that the API naming guidelines could be more clearly written regarding the priorities of these cases. Unless there's further guidance on the choice here, I argue these functions should definitely be named `to_char`, `to_u8`, etc. Similarly, the conversions in `char` should be `char::to_ascii` and `char::to_ascii_unchecked`, both accepting `self` as input instead of `&self`.

**Comment by programmerjake** [comment]
imo for consistency, char literals shouldn't coerce to `ascii::Char`, just like how `'a'` doesn't coerce to `u8` -- there's a separate `b'a'` for that, so I think we should have separate `a'a'` and `a"foo"` literals

**Comment by vinhkhale** [comment]
To avoid confusion between char and ascii::Char, my two cents is to introduce the type as a new primitive and simply name it ascii, a7, or u7. We have a number of types floating around:
'A' char, [char], *"AA" str
b'A' u8, *b"AA" [u8]
I wouldn't mind seeing something like:
a'A' u7, *a"AA" [u7]
Granted we'd want it to be repr'd as a u8, so perhaps:
a'A' ascii, *a"AA" [ascii]
Internally, a str is a [u8] with an additional constraint of being valid UTF8. Here, we have as_bytes and from_utf8. We can imagine that a u7 is just a u8 with the additional constraint that the value is less than 128. Every [u7] is a str, and every str is a [u8]. So we can probably name the conversion functions as_bytes or as_str, assuming this is just a view rather than a conversion, and from_ascii([u8]) or s.to_ascii(), returning Result<[u7], AsciiError>.

**Comment by vinhkhale** [comment]
I can see the point in wanting an AsciiChar type backed by a single byte internally in the same way that a char is 4 bytes internally. What I care about is that there's a type the size of a u8 that is always a valid ASCII character, a type that represents a string of ASCII characters whose length in characters matches its size in bytes, and conversions from the ASCII string type to UTF8 strings are infallible and require no copies or allocations. Whether the invariant of being less than 128 is maintained internally or offbloaded to u7's type guarantees doesn't really matter to me.

That said, I would like to +1 one of the points made in the original issue. We shouldn't name this ascii::Char and instead opt for AsciiChar, AsciiStr, and AsciiString. This kind of naming prevents confusion in code that uses both char and ascii::Char, and it is more in line with other types like CStr and OsString.

**Comment by programmerjake** [comment]
> AsciiStr

that should just be `[AsciiChar]`

**Comment by vinhkhale** [comment]
> we do want different types for printing so we can differentiate between printing as `[a'A', a'B', a'C']` and `a"ABC"`)

A separate AsciiStr or AsciiString type would also be right place to define ASCII-optimized string utilities like to_ascii_uppercase, which we might simply rename to_uppercase, since the fact that it's ASCII is evident from the type.

**Comment by programmerjake** [comment]
> As a small counterexample, I'll note that the standard library does ship both `cmp::Ordering` and `sync::atomic::Ordering`

I'd argue that confusing `ascii::Char` with `char` is a much bigger issue since you very rarely need to use both `cmp::Ordering` and `sync::atomic::Ordering` in the same code, whereas you're quite likely to need to process ASCII and UTF-8 in the same code, which greatly increases the chances of confusion.

**Comment by vinhkhale** [comment]
> I'll note that the standard library does ship both `cmp::Ordering` and `sync::atomic::Ordering`, which coexist (via qualified use or renamed import) without too much issue.

Those types exist in separate problem domains and have different meanings. A better comparison (no pun intended) would be to `std::sync::broadcast::Sender` and `mpsc::Sender`. But even the name clash here is forgivable because they are two flavors of the same tool and are drop-in replacements for each other.

`AsciiChar` and `char` are distinct data types that might need to interoperate in many use cases. It facilitates communication to have separate default names for each type. We have the `AtomicBool` and `bool` distinction for the same reason.

**Comment by vinhkhale** [comment]
> > `std::sync::broadcast::Sender`
> 
> that doesn't exist in `std`
Ah, whoops. `std::sync::mpmc::Sender` then. Regardless, my point still stands.

**Comment by bjoernager** [comment]
IIRC one "issue" with the `Ascii` name was that it could be confused for a string type. But with ASCII, specifically, we can just use `[Ascii]` etc., and I've always preferred the name with that in mind.

**Comment by clarfonthey** [comment]
There *is* a compelling case to name the type something like `AChar` if `AStr` were its own, dedicated type, but since `AStr` is literally identical to `[AChar]` in every way, I think `Ascii` and `[Ascii]` is a bit nicer and clearer.

But, ultimately, I think that we should not go with `ascii::Char`, since importing `ascii::Char` means that `Char` is ASCII and `char` is Unicode and I *hate* that.

**Comment by kupiakos** [comment]
> `Cow` is also an adjective (copy-on-write) but used as a noun. I don't think that it's unreasonable to be flexible with language in this case to avoid being more specific than necessary.

`Cow<T>` is a copy-on-write container for `T`, so arguably `Cow` is still an adjective modifying the noun `T`.

`Ascii` on its own feels like it's missing the `Char` noun part. It reads to me like a trait or a wrapper type (like `Cow`). My preferred names are `ascii::AsciiChar`, `ascii::AChar`, and then `ascii::Ascii`, in that order. I could get used to the abbreviation, though.

None of these have the inevitable readability concerns as `ascii::Char` in diagnostics and 3p rustdoc - that leads me to think the extra typing (or `use as`) is worth it. `sync::Ordering` and `cmp::Ordering` are rarely used in the same contexts, while `char` and this ASCII char would be.

**Comment by vinhkhale** [comment]
> > importing ascii::Char means that Char is ASCII and char is Unicode
> 
> What if you don't import `ascii::Char` and instead import `ascii` and use `ascii::Char` everywhere? It's much clearer than `AChar`, and as close to the previously proposed `AsciiChar` as possible.

The library doesn't ask its consumers to do this with other types of the same flavor, e.g., `CStr` and `AtomicBool`. The name of the type itself should communicate the relevant properties of the type, insofar as it is to be distinguished from other types likely to be used in the same context. The property should not be communicated through the module name. Regardless, I am not sure that importing the whole `ascii` module should necessarily be encouraged when the consumer knows exactly one type is needed.

**Comment by ChrisDenton** [comment]
There are currently three other unresolved questions listed aside from naming.

**Comment by tmke8** [comment]
I thought this was also blocked on switching the implementation from an enum to using [pattern types](https://github.com/rust-lang/rust/issues/123646). I think the currently implemented syntax for those is

```rust
type Char = std::pat::pattern_type!(u8 is 0..128);
```

EDIT: fix the upper bound
